### ISSUE TITLE
Entity storage as class property false positive

### ISSUE THREAD

**Issue by claudiu-cristea** [mglaman/phpstan-drupal#1021]
# Bug report

In a class where we use a lot the entity storage, we've created a protected methods that caches and returns it. It was exactly done in order to avoid injecting the storage in constructor. But now we're getting:

> Storing entity storage as a class property is not recommended. Call Drupal\Core\Entity\EntityTypeManagerInterface::getStorage() at the call-site instead.

### Code snippet that reproduces the problem

```php
  protected function getStorage(): NodeStorageInterface {
    if (!isset($this->storage)) {
      $this->storage = $this->entityTypeManager->getStorage('node');
    }
    return $this->storage;
  }
```

**Comment by rupertj** [comment]
I also get a false positive for this rule when writing a class that extends EntityListBuilder, which I'm doing for a custom entity in one of my modules.

The EntityListBuilder constructor requires you to inject the storage for the entity. Core does this in EntityListBuilder::createInstance(). When I extend EntityListBuilder and override createInstance to add an additional property I want to inject, I have to inject the list builder like core does. This trips the entity storage as class property error.

**Comment by adrian-lorenc** [comment]
The rule is intentional

Entity storage objects can become stale or lose context if cached or stored long-term across different execution states or bootstrap phases.

See https://github.com/drupal/drupal/blob/main/core/lib/Drupal/Core/Entity/EntityTypeManager.php#L44
Every call to getStorage('node') after the first one is just an array lookup
